dali/kernels/imgproc/sampler.h

```diff
@@ -30,165 +30,183 @@ namespace kernels {
 
 struct BorderClamp {};
 
-template <DALIInterpType interp, typename In>
+template <DALIInterpType interp, int spatial_ndim, typename In>
 struct Sampler;
 
-template <DALIInterpType interp, typename T>
-DALI_HOST_DEV Sampler<interp, T> make_sampler(const Surface2D<T> &surface) {
-  return Sampler<interp, T>(surface);
+template <DALIInterpType interp, typename In>
+using Sampler2D = Sampler<interp, 2, In>;
+
+template <DALIInterpType interp, typename In>
+using Sampler3D = Sampler<interp, 3, In>;
+
+
+template <DALIInterpType interp, int ndim, typename T>
+DALI_HOST_DEV
+Sampler<interp, ndim, std::remove_const_t<T>> make_sampler(const Surface<ndim, T> &surface) {
+  return Sampler<interp, ndim, std::remove_const_t<T>>(surface);
 }
 
 template <typename T>
-DALI_HOST_DEV T GetBorderChannel(const T *values, int c) {
+DALI_HOST_DEV DALI_FORCEINLINE T GetBorderChannel(const T *values, int c) {
   return values[c];
 }
 
 template <typename T>
-DALI_HOST_DEV constexpr
+DALI_HOST_DEV DALI_FORCEINLINE constexpr
 enable_if_t<!std::is_pointer<T>::value, T> GetBorderChannel(const T &value, int) {
   return value;
 }
 
-template <typename In>
-struct Sampler<DALI_INTERP_NN, In> {
+/**
+ * @brief Checks if all coordinates in `coords` are positive and below the respective
+ *        coordinate in `limit`.
+ *
+ * @remarks `limit` is expected to have non-negative coordinates - otherwise, the result
+ *          is undefined.
+ */
+template <int n>
+DALI_HOST_DEV DALI_FORCEINLINE
+constexpr bool all_in_range(ivec<n> coords, ivec<n> limit) {
+  for (int i = 0; i < n; i++) {
+    if (static_cast<unsigned>(coords[i]) >= static_cast<unsigned>(limit[i]))
```

Comment:
At first glance I don't understand why you are casting to unsigned. I'd rather do the explicit comparison between `0` and `limit[i]` or at least add a comment saying why you are casting to unsigned